File: Java/com/leetcode/uniquePathsII/UniquePathsII.java
package com.leetcode.uniquePathsII;

public class UniquePathsII {
	public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid == null || obstacleGrid[0][0] == 1) {
            return 0;
        }
        
        int[][] state = new int[obstacleGrid.length][obstacleGrid[0].length];
        
        for (int i = 0; i < obstacleGrid.length; i++) {
            for (int j = 0; j < obstacleGrid[0].length; j++) {
                
                if (obstacleGrid[i][j] == 1) {
                    state[i][j] = 0;
                    continue;
                }
                
                if (i - 1 >= 0 && j - 1 < 0) {
                    state[i][j] = state[i - 1][j];
                } else if (i - 1 < 0 && j - 1 >= 0) {
                    state[i][j] = state[i][j - 1];
                } else if (i - 1 < 0 && j - 1 < 0) {
                    state[i][j] = 1;
                } else {
                    state[i][j] = state[i - 1][j] + state[i][j - 1];
                }
                
            }
        }
        
        return state[obstacleGrid.length - 1][obstacleGrid[0].length - 1];
    }
}

File: Python/unique_paths_ii/unique_paths_ii.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        if len(obstacleGrid) == 0 or \
                len(obstacleGrid[0]) == 0 or \
                obstacleGrid[0][0] == 1:
            return 0
        obstacleGrid[0][0] = 1
        for i in range(1, len(obstacleGrid)):
            if obstacleGrid[i][0] == 1:
                obstacleGrid[i][0] = 0
            else:
                obstacleGrid[i][0] = obstacleGrid[i-1][0]

        for j in range(1, len(obstacleGrid[0])):
            if obstacleGrid[0][j] == 1:
                obstacleGrid[0][j] = 0
            else:
                obstacleGrid[0][j] = obstacleGrid[0][j-1]

        for i in range(1, len(obstacleGrid)):
            for j in range(1, len(obstacleGrid[i])):
                if obstacleGrid[i][j] == 1:
                    obstacleGrid[i][j] = 0
                else:
                    obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1]

        return obstacleGrid[-1][-1]
